### PR TITLE
First commit to the golang implementation.

### PR DESCRIPTION
**Pull request by kklipsch**
This gets the code to the point where a simple add/stat query works.  Test coverage is not great but should be easy to add.

### CODE REVIEW

**Comment by kklipsch** [comment]
Hmm.  Interesting, if you comment on the specific commit it doesnt hide the comments in teh conversation when a later change causes it to be no longer valid.

File: add.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/kklipsch/cli"
+)
+
+func Add(c *cli.Context, kayDir KayDir, index *index) error {
```
Comment (bhand-mm) [inline add.go]: In general you want to check pointers for nil, catch the panic (here or in a caller) or expect that the app should just die if any of these are unexpectedly nil.  That last one is fine but its something to keep in mind.

File: add.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+	year, year_parsed, err := ParseYear(file)
+	if err != nil {
+		return err
+	}
+
+	if !year_parsed {
+		return fmt.Errorf("No year provided or parsed!")
+	}
+
+	index.CreateRecord(year, file, Note(""))
```
Comment (bhand-mm) [inline add.go:20]: You are throwing away the return values here.  One of them is an error.

File: autocomplete/kay.bash [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+     local cur prev opts base
+     COMPREPLY=()
+     cur="${COMP_WORDS[COMP_CWORD]}"
+     prev="${COMP_WORDS[COMP_CWORD-1]}"
+     opts=$( ${COMP_WORDS[@]:0:$COMP_CWORD} --generate-bash-completion )
+     COMPREPLY=( $(compgen -W "${opts}" -- ${cur}) )
+     return 0
+ }
+  
+ complete -F _cli_bash_autocomplete $PROG
```
Comment (bhand-mm) [inline autocomplete/kay.bash:15]: Fancy!

File: file.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+)
+
+func GetFilesFromDir(path string) ([]File, error) {
+	dir, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return getFilesFromList(path, dir), nil
+}
```
Comment (bhand-mm) [inline file.go:15]: oh would that go had "map"!

File: file.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+func getFilesFromList(path string, content []os.FileInfo) []File {
+	files := []File{}
+	for _, f := range content {
+		if !f.IsDir() {
+			files = append(files, File(f.Name()))
+		}
+	}
+	return files
+}
```
Comment (bhand-mm) [inline file.go:26]: and "filter"!

File: file_test.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
+package main
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+)
+
+func TestGetFilesFromDir(t *testing.T) {
```
Comment (bhand-mm) [inline file_test.go]: Based on what this is testing perhaps call it "TestGetOnlyFilesFromDir"?

File: index.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+	"fmt"
+	"time"
+)
+
+type Year int
+type File string
+type Tag string
+type Note string
+
+type index struct {
```
Comment (bhand-mm) [inline index.go]: I'm pretty sure this could just be a type alias too.

File: index.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+func (i *index) Records() []*record {
+	records := []*record{}
+	for _, record := range i.records {
+		records = append(records, record)
+	}
+	return records
+}
+
+type record struct {
+	//changing the file should be done through the index
```
Comment (bhand-mm) [inline index.go:78]: An excellent time for a comment.  I would say this should be in its own package though.  Making it lowercase is less meaningful unless there are package boundaries.

File: index.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+	LastWritten time.Time
+	Tags        []Tag
+}
+
+func (r record) File() File {
+	return r.file
+}
+
+func (r record) HasTag(tag Tag) bool {
+	//golang doesnt have a contains function for arrays?
```
Comment (bhand-mm) [inline index.go:93]: You are _lucky_ it even has append/insert.  Also this is a slice not an array.  A distinction of which we must be ever vigilant!

File: index_test.go [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+func TestCreateRecordForSameFileTwice(t *testing.T) {
+	i := EmptyIndex()
+	f := File("foo")
+
+	i.CreateRecord(Year(1952), f, "")
+	_, err := i.CreateRecord(Year(1953), f, "")
+	FailIfNoError(t, err, "Should not be able to add file twice")
+}
+
+func TestGetRecordForFileFile(t *testing.T) {
```
Comment (bhand-mm) [inline index_test.go]: FileFile?

File: init.go [hunk 11]
```diff
@@ -0,0 +1,12 @@
+package main
+
+import "github.com/kklipsch/cli"
+
+func Initialize(c *cli.Context) error {
```
Comment (bhand-mm) [inline init.go:5]: Its worth noting that Go actually has an "init" entry point.  Each package may define an init function to be called exactly once if its imported.  That is the reason you see imports that start with an underscore (import _ "somepackage").  That is, to paraphrase, "importing for the side effects only".

I say all this to make for two reasons:
1. to make you cringe.
2. so you are aware and don't define something that doesn't work like you expect.  This function is fine.  It is specifically "func init() {}" that you have to watch out for.

http://golang.org/ref/spec#Package_initialization

File: kaydir.go [hunk 12]
```diff
@@ -0,0 +1,36 @@
+package main
+
+import (
+	"os"
+	"path/filepath"
+)
+
+const kay_meta_dir string = ".kay"
```
Comment (bhand-mm) [inline kaydir.go]: _?

File: main.go [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+func main() {
+	err := NewKay().Run(os.Args)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+
+	os.Exit(0)
+}
```
Comment (bhand-mm) [inline main.go:17]: For some reason this seems more "you":

```
func main() {
    err := NewKay().Run(os.Args)

    exitCode := 0
    if err != nil {
        fmt.Fprintln(os.Stderr, err.Error())
        exitCode = 1
    }

    os.Exit(exitCode)
}
```

But perhaps instead just:

```
func main() {
    err := NewKay().Run(os.Args)

    if err != nil {
        log.Fatal(err) // "log" uses STDERR by default
    }
}
```

http://golang.org/pkg/log/#Fatal

File: main.go [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+			Action: KayBased(Stat),
+		},
+		{
+			Name:   "init",
+			Usage:  "initialize a new kay directory",
+			Action: Initialize,
+		},
+	}
+
+	return app
```
Comment (bhand-mm) [inline main.go:97]: Fancy!

File: serialization.go [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+func BuildIndex(indexDir string) (*index, error) {
+	if err := os.MkdirAll(indexDir, 0755); err != nil {
```
Comment (bhand-mm) [inline serialization.go:14]: I think you already know this but be careful of the scope of the "if ;" statement.

File: serialization.go [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+func readRecord(file File, path string, index *index) (*record, error) {
+	bytes, readErr := ioutil.ReadFile(filepath.Join(path, string(file)))
+	if readErr != nil {
+		return nil, readErr
+	}
+
+	return DeserializeRecord(file, index, bytes)
+}
+
+func readAllRecords(index *index, path string, files []File) (*index, error) {
```
Comment (bhand-mm) [inline serialization.go]: This kinda feels like a method of *index

File: serialization.go [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+	return index, nil
+}
+
+func createTempIndexDir(indexDir string) (string, error) {
+	escapedName := strings.Replace(indexDir, "/", ".", -1)
+	tempdir, temperr := ioutil.TempDir("", escapedName)
+	if temperr != nil {
+		return "", temperr
+	} else {
+		return tempdir, nil
```
Comment (bhand-mm) [inline serialization.go:89]: On some versions of Go i think this throws an error that there is no return at the end of the function (even though we both know there is).

**Pull request merged**